File: backend/src/main/java/com/expenseguard/recurring/entity/RecurringTransaction.java
package com.expenseguard.recurring.entity;

import com.expenseguard.auth.entity.User;
import com.expenseguard.transaction.entity.Account;
import com.expenseguard.transaction.entity.Category;
import com.expenseguard.transaction.entity.TransactionType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * Domain Entity representing a template for recurring financial transactions.
 */
@Entity
@Table(
        name = "recurring_transactions",
        indexes = {
                @Index(name = "idx_rec_tx_user_id", columnList = "user_id"),
                @Index(name = "idx_rec_tx_next_run_date", columnList = "next_run_date"),
                @Index(name = "idx_rec_tx_active_next_run", columnList = "active, next_run_date")
        }
)
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class RecurringTransaction {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private TransactionType type;

    @Column(name = "amount", nullable = false, precision = 19, scale = 4)
    private BigDecimal amount;

    @Column(name = "description")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "frequency", nullable = false)
    private RecurrenceFrequency frequency;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "next_run_date", nullable = false)
    private LocalDate nextRunDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Builder.Default
    @Column(name = "active", nullable = false)
    private boolean active = true;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        Instant now = Instant.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }
}

File: mobile/src/types/recurringTransaction.ts
import { TransactionType } from './transaction';

export type RecurrenceFrequency = 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY';

export interface RecurringTransactionRequest {
  accountId: string;
  categoryId: string;
  type: TransactionType;
  amount: number;
  description?: string;
  frequency: RecurrenceFrequency;
  startDate: string; // YYYY-MM-DD
  nextRunDate: string; // YYYY-MM-DD
  endDate?: string; // YYYY-MM-DD
}

export interface RecurringTransactionResponse {
  id: string;
  accountId: string;
  categoryId: string;
  categoryName: string;
  type: TransactionType;
  amount: number;
  description?: string;
  frequency: RecurrenceFrequency;
  startDate: string;
  nextRunDate: string;
  endDate?: string;
  active: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface RecurringTransactionStatusRequest {
  active: boolean;
}

export interface RecurringExecutionResponse {
  message: string;
  executedCount: number;
}
